src/StatsTools/variational_energy_estimator.jl

```diff
@@ -0,0 +1,82 @@
+"""
+    variational_energy_estimator(shifts, overlaps; skip=0, kwargs...)
+    variational_energy_estimator(df::DataFrame; max_replicas=:all, skip=0, kwargs...)
+    -> r::RatioBlockingResult
+Compute the variational energy estimator from the replica time series of the `shifts` and
+coefficient vector `overlaps` by blocking analysis.
+The keyword argument `max_replicas` can be used to constrain the number of replicas
+processed to be smaller than all available in `df`.
+Other keyword arguments are passed on to [`ratio_of_means()`](@ref).
+Returns a [`RatioBlockingResult`](@ref).
+
+An estimator for the variational energy
+```math
+\\frac{⟨\\mathbf{c}⟩^† \\mathbf{H}⟨\\mathbf{c}⟩}{⟨\\mathbf{c}⟩^†⟨\\mathbf{c}⟩}
+```
+is calculated from
+```math
+Ē_{v}  =  \\frac{\\sum_{a<b}^R \\overline{(S_a+S_b) \\mathbf{c}_a^† \\mathbf{c}_b}}
+               {2\\sum_{a<b}^R \\overline{\\mathbf{c}_a^† \\mathbf{c}_b}} ,
+```
+where the sum goes over distinct pairs out of the ``R`` replicas. See
+[arXiv:2103.07800](http://arxiv.org/abs/2103.07800).
+
+The `DataFrame` version can extract the relevant information from the result of
+[`lomc!`](@ref). Set up [`lomc!`](@ref) with the keyword argument
+`replica = AllOverlaps(R)` and `R ≥ 2`.
+
+See [`AllOverlaps`](@ref).
+"""
+function variational_energy_estimator(shifts, overlaps; kwargs...)
+    num_replicas = length(shifts)
+    if length(overlaps) ≠ binomial(num_replicas, 2)
+        throw(ArgumentError(
+            "The number of `overlaps` needs to be `binomial(length(shifts),2)`."
+        ))
+    end
+    denominator = sum(overlaps)
+    numerator = zero(shifts[1])
+    count_overlaps = 0
+    for i in 1:num_replicas, j in i+1:num_replicas
+        count_overlaps += 1
+        @. numerator += 1 / 2 * (shifts[i] + shifts[j]) * overlaps[count_overlaps]
+    end
+    return ratio_of_means(numerator, denominator; kwargs...)
+end
+
+function variational_energy_estimator(df::DataFrame; max_replicas=:all, kwargs...)
+    if "shift" in names(df)
+        throw(ArgumentError(
+            "`DataFrame` looks like a non-initiator output. Use keyword \
+            `replica=AllOverlaps(n)` with n≥2 in `lomc!()` to set up replicas!"
+        ))
+    end
+    num_replicas = length(filter(startswith("norm_"), names(df))) # number of replicas
+    if iszero(num_replicas)
+        throw(ArgumentError(
+            "No replicas found. Use keyword \
+            `replica=AllOverlaps(n)` with n≥2 in `lomc!()` to set up replicas!"
+        ))
+    end
+    @assert num_replicas ≥ 2 "At least two replicas are needed, found $num_replicas"
```

Comment:
Should these conditions be merged with the one that looks for `shift`? Looks like we are basically checking the same thing three times.

Reply:
Good point. I deleted the first check as this is fully covered by the second. I'd leave the @assert statements in, as they don't cost anything. They should never be triggered but still protect from possible coding bugs.